packages/flutter_tools/test/widget_preview_scaffold.shard/widget_preview_scaffold/test/dtd_services_test.dart

```diff
@@ -0,0 +1,133 @@
+
+  tearDown(() async {
+    await dtdServer.shutdownHooks.runShutdownHooks(logger);
+  });
+
+  group('$WidgetPreviewDtdServices', () {
+    testUsingContext(
+      'handles ${WidgetPreviewDtdServices.kHotRestartPreviewer} invocations',
+      () async {
+        // Start DTD and register the widget preview DTD services with a custom handler for hot
+        // restart requests.
+        final hotRestartRequestCompleter = Completer<void>();
+        dtdServer = WidgetPreviewDtdServices(
+          fs: MemoryFileSystem.test(),
+          logger: logger,
+          shutdownHooks: ShutdownHooks(),
+          dtdLauncher: DtdLauncher(
+            logger: logger,
+            artifacts: globals.artifacts!,
+            processManager: globals.processManager,
+          ),
+          onHotRestartPreviewerRequest: hotRestartRequestCompleter.complete,
+          project: FakeFlutterProject(),
+        );
+        await dtdServer.launchAndConnect();
+
+        // Connect to the DTD instance and invoke the hot restart endpoint.
+        final dtd = WidgetPreviewScaffoldDtdServices();
+        await dtd.connect(dtdUri: dtdServer.dtdUri);
+
+        await dtd.hotRestartPreviewer();
+
+        // Ensure the custom handler is actually invoked.
+        await hotRestartRequestCompleter.future;
+      },
+      overrides: <Type, Generator>{ProcessManager: () => loggingProcessManager},
+    );
+
+    testUsingContext(
+      'can set and retreive values from $PersistentPreferences',
+      () async {
+        dtdServer = WidgetPreviewDtdServices(
+          fs: MemoryFileSystem.test(),
+          logger: logger,
+          shutdownHooks: ShutdownHooks(),
+          dtdLauncher: DtdLauncher(
+            logger: logger,
+            artifacts: globals.artifacts!,
+            processManager: globals.processManager,
+          ),
+          onHotRestartPreviewerRequest: () {},
+          project: FakeFlutterProject(),
+        );
+        await dtdServer.launchAndConnect();
+
+        // The properties file should be created by the PersistentProperties constructor.
+        final File preferencesFile = dtdServer.preferences.file;
+        expect(preferencesFile.existsSync(), true);
+
+        // Connect to the DTD instance.
+        final dtd = WidgetPreviewScaffoldDtdServices();
+        await dtd.connect(dtdUri: dtdServer.dtdUri);
+
+        const kTestKey = 'myKey';
+
+        // The preferences file should be empty.
+        expect(await dtd.getPreference(kTestKey), null);
+        expect(preferencesFile.readAsStringSync(), isEmpty);
+
+        // Set a preference and ensure it's read back.
+        const kFirstValue = 'foo';
+        await dtd.setPreference(kTestKey, kFirstValue);
+        expect(await dtd.getPreference(kTestKey), kFirstValue);
+        expect(json.decode(preferencesFile.readAsStringSync()), {
+          kTestKey: kFirstValue,
+        });
+
+        // Overwrite kTestKey and ensure it's read back.
+        const kSecondValue = 'bar';
```

Comment:
Is it worth having a test that removes the key (since I think there was is special handling for nulls/missing keys?)